examples/rust-data-loader/rust/src/lib.rs

```diff
@@ -0,0 +1,268 @@
+    channels: BTreeSet<u16>,
+}
+
+impl NDJsonIterator {
+    fn open(rows: Rc<RefCell<Vec<Row>>>, args: &MessageIteratorArgs) -> Self {
+        Self {
+            rows: rows.clone(),
+            index: 0.into(),
+            start: args.start_time.unwrap_or(0),
+            end: args.end_time.unwrap_or(u64::MAX),
+            channels: args.channels.iter().copied().collect(),
+        }
+    }
+}
+
+impl MessageIterator for NDJsonIterator {
+    type Error = anyhow::Error;
+
+    fn next(&self) -> Option<Result<Message, Self::Error>> {
+        loop {
+            let index = *self.index.borrow();
+            self.index.replace(index + 1);
+            let rows = self.rows.borrow();
+            let row = rows.get(index);
+            if let Some(time) = row.map(|r| seconds_to_nanos(r.get_time())) {
+                if time < self.start {
+                    continue;
+                }
+                if time > self.end {
+                    return None;
+                }
+            };
+            match row {
+                None => return None,
+                Some(Row::Accelerometer(accel)) => {
+                    if self.channels.contains(&ACCELEROMETER_CHANNEL_ID) {
+                        return Some(Ok(accel.into()));
+                    }
+                }
+                Some(Row::Temperature(temperature)) => {
+                    if self.channels.contains(&TEMPERATURE_CHANNEL_ID) {
+                        return Some(Ok(temperature.into()));
+                    }
+                }
+            };
+        }
+    }
+}
+
+// floating point time in seconds to u64 nanoseconds
+fn seconds_to_nanos(time_seconds: f64) -> u64 {
+    (time_seconds * 1.0e9) as u64
+}
+
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+#[serde(tag = "type")]
+enum Row {
+    #[serde(rename = "accelerometer")]
+    Accelerometer(Accelerometer),
+    #[serde(rename = "temperature")]
+    Temperature(Temperature),
+}
+
+impl Row {
+    fn get_time(&self) -> f64 {
+        match self {
+            Row::Accelerometer(accel) => accel.time,
+            Row::Temperature(temperature) => temperature.time,
+        }
+    }
+}
+
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+struct Accelerometer {
+    time: f64, // in seconds
+    x: f64,
+    y: f64,
+    z: f64,
+}
+
+#[derive(Debug, Clone, foxglove::Encode, serde::Serialize, serde::Deserialize)]
+struct Temperature {
+    time: f64, // in seconds
+    ambient: f64,
+    cpu0: f64,
+    cpu1: f64,
+    cpu2: f64,
+    cpu3: f64,
+}
+
+impl From<&Accelerometer> for Message {
+    fn from(accel: &Accelerometer) -> Message {
+        let time_nanos = seconds_to_nanos(accel.time);
+        Message {
+            channel_id: ACCELEROMETER_CHANNEL_ID,
+            log_time: time_nanos,
+            publish_time: time_nanos,
+            data: Vector3 {
+                x: accel.x,
+                y: accel.y,
+                z: accel.z,
+            }.encode_to_vec(),
```

Comment:
Where is the `time` field from the `Accelerometer` struct defined above? Maybe you don't have to define that struct at all if you are using `Vector3` ?